### ISSUE TITLE
Broker address changes after disconnecting and reconnecting brokers in an embedded kafka cluster

### ISSUE THREAD

**Issue by sunnylohani** [mguenther/kafka-junit#48]
```
private EmbeddedKafkaCluster cluster = provisionWith(
  EmbeddedKafkaClusterConfig.create()
        .provisionWith(EmbeddedKafkaConfig.create()
            .withNumberOfBrokers(3)
            .with(KafkaConfig$.MODULE$.NumPartitionsProp(), "3")
            .with(KafkaConfig$.MODULE$.DefaultReplicationFactorProp(), "3")
            .with(KafkaConfig$.MODULE$.MinInSyncReplicasProp(), "2")
            .with(KafkaConfig$.MODULE$.OffsetsTopicReplicationFactorProp(), "3")
            .build())
        .build()
);

cluster.start();
log.info("Starting Kafka cluster with brokers: {}", cluster.getBrokerList());

// Disconnect until the remaining number of brokers fall below the minimum ISR size
Set<Integer> disconnectedBrokers = cluster.disconnectUntilIsrFallsBelowMinimumSize(TOPIC_NAME);
delay(5);
log.info("Kafka cluster after disconnect: {}", cluster.getBrokerList());

// Reconnect brokers
cluster.connect(disconnectedBrokers);
delay(5);
log.info("Kafka cluster after reconnect: {}", cluster.getBrokerList());
```

**OUTPUT:**
_Starting Kafka cluster with brokers: localhost:57839,localhost:57842,localhost:57845
Kafka cluster after disconnect: localhost:57845
Kafka cluster after reconnect: localhost:58001,localhost:58005,localhost:57845_

Is it possible to retain the same broker address after reconnect ? So that the kafka producers created earlier resume to function normally.

**Comment by sunnylohani** [comment]
Can I get an answer to this issue ?

**Comment by mguenther** [comment]
Thanks for pointing out the issue!

Unfortunately, this is not possible with the latest release of kafka-junit. If you start an embedded Kafka cluster with multiple brokers, all of those brokers share the same configuration. The broker configuration for `KafkaConfig$.MODULE$.PortProp()` defaults to `0`, which works in this case, because the broker selects an ephemeral port to bind its listening socket to. Overriding this value only works in the case where you have a single broker instance, otherwise all other broker instances would try to assign their listening socket to the same port. But this defeats the purpose of your test case anyhow.

To solve this, the lifecycle of an `EmbeddedKafka` broker must retain the server port and override `KafkaConfig$.MODULE$.PortProp()` on a restart. This is currently not supported by kafka-junit, but nonetheless should be supported.

I'll have to look into this issue.

**Comment by mguenther** [comment]
I've tested the approach and it seems to work just fine. I do have to flesh this a bit out, because the simplicity of kafka-junit's configuration must be retained and with the fix there is the possibility to configure the embedded cluster in a way that it won't work as expected. This will take some time to address properly. I'll update the issue as soon as this is fixed.

**Comment by sunnylohani** [comment]
Thanks a lot for detailed info on the issue. Will wait for the fix.

**Comment by mguenther** [comment]
Branch issue-48 has been merged back into development. The bugfix will target on-going development of 2.6.0 and is released for the 2.5.x line as version 2.5.1. Release artifacts for 2.5.1 have been uploaded to Maven Central. This will take some time though (up to a couple of hours) until it is available. I'll file a release note for 2.5.1 as soon as it is publicly available.

**Comment by sunnylohani** [comment]
Thanks for the quick fix. Really appreciate! Waiting for artifacts to be available on mvn central.
